From 76197885b5fba5ef1a074bc0f602a8a2e226c6a1 Mon Sep 17 00:00:00 2001
From: Ajay Kumar Gupta <ajay.gupta@ti.com>
Date: Fri, 21 Oct 2011 10:32:12 +0530
Subject: [PATCH 0605/1092] usb: ether_gadget: fix insmod panic after fsg
 rmmod

Fixes kernel panic seen during g_ether.ko module insertion performed
after the FSG gadget rmmod.

Ether gadget doesn't reset the usb_ep(s) during initialization and
so was getting the previous gadget's usb_ep->desc data.

Signed-off-by: Vaibhav Hiremath <hvaibhav@ti.com>
---
 drivers/usb/gadget/epautoconf.c | 1 +
 drivers/usb/gadget/f_ecm.c      | 1 +
 drivers/usb/gadget/f_rndis.c    | 1 +
 drivers/usb/gadget/f_subset.c   | 1 +
 4 files changed, 4 insertions(+)

diff --git a/drivers/usb/gadget/epautoconf.c b/drivers/usb/gadget/epautoconf.c
index 753aa06..a9f58da 100644
--- a/drivers/usb/gadget/epautoconf.c
+++ b/drivers/usb/gadget/epautoconf.c
@@ -380,6 +380,7 @@ void usb_ep_autoconfig_reset (struct usb_gadget *gadget)
 
 	list_for_each_entry (ep, &gadget->ep_list, ep_list) {
 		ep->driver_data = NULL;
+		ep->desc = NULL;
 	}
 #ifdef	MANY_ENDPOINTS
 	in_epnum = 0;
diff --git a/drivers/usb/gadget/f_ecm.c b/drivers/usb/gadget/f_ecm.c
index 11c07cb..58d9172 100644
--- a/drivers/usb/gadget/f_ecm.c
+++ b/drivers/usb/gadget/f_ecm.c
@@ -690,6 +690,7 @@ ecm_bind(struct usb_configuration *c, struct usb_function *f)
 	status = -ENODEV;
 
 	/* allocate instance-specific endpoints */
+	usb_ep_autoconfig_reset(cdev->gadget);
 	ep = usb_ep_autoconfig(cdev->gadget, &fs_ecm_in_desc);
 	if (!ep)
 		goto fail;
diff --git a/drivers/usb/gadget/f_rndis.c b/drivers/usb/gadget/f_rndis.c
index b6e6dac..6614490 100644
--- a/drivers/usb/gadget/f_rndis.c
+++ b/drivers/usb/gadget/f_rndis.c
@@ -684,6 +684,7 @@ rndis_bind(struct usb_configuration *c, struct usb_function *f)
 	status = -ENODEV;
 
 	/* allocate instance-specific endpoints */
+	usb_ep_autoconfig_reset(cdev->gadget);
 	ep = usb_ep_autoconfig(cdev->gadget, &fs_in_desc);
 	if (!ep)
 		goto fail;
diff --git a/drivers/usb/gadget/f_subset.c b/drivers/usb/gadget/f_subset.c
index 21ab474..160ba02 100644
--- a/drivers/usb/gadget/f_subset.c
+++ b/drivers/usb/gadget/f_subset.c
@@ -307,6 +307,7 @@ geth_bind(struct usb_configuration *c, struct usb_function *f)
 	status = -ENODEV;
 
 	/* allocate instance-specific endpoints */
+	usb_ep_autoconfig_reset(cdev->gadget);
 	ep = usb_ep_autoconfig(cdev->gadget, &fs_subset_in_desc);
 	if (!ep)
 		goto fail;
-- 
1.7.11.2

